pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,310 @@
+// the span that is recorded for the new key is returned, as it may be a subspan
+// of the span passed in for the key.
+//
+// Flush should be called after the last key is written to ensure that the SST
+// is written to the destination.
+func (s *SSTSinkKeyWriter) WriteKey(ctx context.Context, key storage.MVCCKey, value []byte) error {
+	if len(s.flushedFiles) == 0 {
+		return errors.AssertionFailedf(
+			"no BackupManifest_File to write key to, call Reset before WriteKey",
+		)
+	}
+	if err := s.maybeFlushOOOKey(ctx, key.Key); err != nil {
+		return err
+	}
+	s.setMidRowForPrevKey(key.Key)
+	err := s.maybeDoSizeFlushWithNextKey(ctx, key.Key)
+	if err != nil {
+		return err
+	}
+
+	// At this point, because of the Reset invariant, we can make the following
+	// assumptions about the new key:
+	// - The new key comes after the previous key.
+	// - This new key belongs to the span of the last BackupManifest_File in
+	//   s.flushedFiles.
+	if len(s.flushedFiles) == 0 {
+		panic(
+			"unexpected missing BackupManifest_File after flushes when writing key",
+		)
+	}
+	elidedKey, prefix, err := elideMVCCKeyPrefix(key, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	s.elidedPrefix = append(s.elidedPrefix[:0], prefix...)
+	if err := s.sst.PutRawMVCC(elidedKey, value); err != nil {
+		return err
+	}
+
+	// TODO (kev-cao): Should add some method of tracking `Rows` (and potentially
+	// `IndexEntries`).
+	keyAsRowCount := roachpb.RowCount{
+		DataSize: int64(len(key.Key)) + int64(len(value)),
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
+	if lastFile.EntryCounts.DataSize >= fileSpanByteLimit {
+		// As we are splitting due to the fileSpanByteLimit, the new span will need
+		// to contiguously extend the last span. We can use the current key as an
+		// exclusive end key for the last file and an inclusive start key for the
+		// new file.
+		newSpan := roachpb.Span{
+			Key:    key.Key,
+			EndKey: lastFile.Span.EndKey,
+		}
+		lastFile.Span.EndKey = newSpan.Key
+		if err := s.Reset(ctx, newSpan); err != nil {
+			return err
+		}
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	lastFile.EntryCounts.Add(keyAsRowCount)
+	s.flushedSize += keyAsRowCount.DataSize
+	s.prevKey = append(s.prevKey[:0], key.Key...)
+
+	return nil
+}
+
+// Reset resets the SSTSinkKeyWriter to write to a new span. It ensures that an
+// open BackupManifest_File exists for the new span.
+// Any time a new span is being written, Reset MUST be called prior to any
+// WriteKey calls.
+func (s *SSTSinkKeyWriter) Reset(ctx context.Context, newSpan roachpb.Span) error {
+	if err := s.maybeFlushOOOSpans(ctx, newSpan); err != nil {
+		return err
+	}
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	var isContiguous, isSame bool
```

Comment:
when would `isSame` ever occur?

Reply:
At the time of writing this, it was from me migrating over the logic. In hindsight, I think perhaps we should cause a flush regardless? If someone is calling `Reset` on the same span, then I think it's safe to say they want to reset the `BackupManifest_File` for it, aka create a new one.

Reply:
Now that i think more about this, this will never happen, and it is not really the sink's responsibility to check this. perhaps we should note down a list of invariants required for `restoreSpanEntries` and check them after generateAndSendImport spans.